azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -7,505 +7,1407 @@
+    def test_pipeline_configuration(self, auth_provider):
+        pipeline = IoTHubPipeline(auth_provider)
+        curr_stage = pipeline._pipeline
+
+        expected_stage_order = [
+            pipeline_stages_base.PipelineRootStage,
+            pipeline_stages_iothub.UseAuthProviderStage,
+            pipeline_stages_iothub.HandleTwinOperationsStage,
+            pipeline_stages_base.CoordinateRequestAndResponseStage,
+            pipeline_stages_base.EnsureConnectionStage,
+            pipeline_stages_iothub_mqtt.IoTHubMQTTConverterStage,
+            pipeline_stages_mqtt.MQTTClientStage,
+        ]
+
+        # Assert that all PipelineStages are there, and they are in the right order
+        for i in range(len(expected_stage_order)):
+            expected_stage = expected_stage_order[i]
+            assert isinstance(curr_stage, expected_stage)
+            curr_stage = curr_stage.next
+
+        # Assert there are no more additional stages
+        assert curr_stage is None
+
+    # TODO: revist these tests after auth revision
+    # They are too tied to auth types (and there's too much variance in auths to effectively test)
+    # Ideally IoTHubPipeline is entirely insulated from any auth differential logic (and module/device distinctions)
+    # In the meantime, we are using a device auth with connection string to stand in for generic SAS auth
+    # and device auth with X509 certs to stand in for generic X509 auth
+    @pytest.mark.it(
+        "Runs a SetAuthProviderOperation with the provided AuthenticationProvider on the pipeline, if using SAS based authentication"
+    )
+    def test_sas_auth(self, mocker, auth_provider, device_connection_string):
+        mocker.patch.object(pipeline_stages_base, "PipelineRootStage")
+        auth_provider = SymmetricKeyAuthenticationProvider.parse(device_connection_string)
+        pipeline = IoTHubPipeline(auth_provider)
+        op = pipeline._pipeline.run_op.call_args[0][0]
+        assert pipeline._pipeline.run_op.call_count == 1
+        assert isinstance(op, pipeline_ops_iothub.SetAuthProviderOperation)
+        assert op.auth_provider is auth_provider
+
+    @pytest.mark.it(
+        "Propagates exceptions that occured in execution upon unsuccessful completion of the SetAuthProviderOperation"
+    )
+    def test_sas_auth_op_fail(self, mocker, auth_provider, device_connection_string):
+        mocker.patch.object(pipeline_stages_base, "PipelineRootStage")
+        auth_provider = SymmetricKeyAuthenticationProvider.parse(device_connection_string)
+        pipeline = IoTHubPipeline(auth_provider)
+        op = pipeline._pipeline.run_op.call_args[0][0]
+        op.error = Exception()
+
+        with pytest.raises(Exception):
+            op.callback(op)
+
+    @pytest.mark.it(
+        "Runs a SetX509AuthProviderOperation with the provided AuthenticationProvider on the pipeline, if using SAS based authentication"
+    )
+    def test_cert_auth(self, mocker, auth_provider, x509):
```

Comment:
>auth_provider [](start = 37, length = 13)

is the fixture getting utilized here ?

Reply:
Not anymore. Good catch.

---
In reply to: [299253973](https://github.com/Azure/azure-iot-sdk-python-preview/pull/120#discussion_r299253973) [](ancestors = 299253973)